openwisp_monitoring/device/apps.py

```diff
@@ -73,3 +77,63 @@ def trigger_device_recovery_checks(cls, instance, **kwargs):
 
         if cache.get(get_device_recovery_cache_key(device=instance)):
             trigger_device_recovery.delay(pk=instance.pk)
+
+    @classmethod
+    def connect_is_working_changed(self):
+        from openwisp_controller.connection.models import DeviceConnection
+        from openwisp_controller.connection.signals import is_working_changed
+
+        is_working_changed.connect(
+            self.is_working_changed_receiver,
+            sender=DeviceConnection,
+            dispatch_uid='change_health_status',
+        )
+
+    @classmethod
+    def is_working_changed_receiver(cls, instance, is_working, **kwargs):
+        from ..check.models import Check
+        from .tasks import trigger_device_recovery
+
+        device = instance.device
+        device_monitoring = device.monitoring
+        initial_status = device_monitoring.status
+        status = 'ok' if is_working else 'problem'
+        if not is_working:
+            device_monitoring.update_status(status)
+            # Create a related notification explaining why it's not working
+            desc = instance.failure_reason
+            opts = dict(
+                actor=device, level='warning', verb='not working', description=desc
+            )
+            if initial_status == 'ok':
+                trigger_device_recovery.delay(pk=device.pk)
+        else:
+            # create a notification that device is working
+            opts = dict(actor=device, level='info', verb='connected successfully')
+            try:
+                # if checks exist trigger them else, set status as 'ok'
+                Check.objects.get(object_id=instance.device.pk)
```

Comment:
if you only need to check whether an element is present, use exists() which is more efficient because designed for this purpose: https://docs.djangoproject.com/en/3.0/ref/models/querysets/#django.db.models.query.QuerySet.exists